### PR TITLE
Option critic network class

### PR DESCRIPTION
**Pull request by emeldar**
This is the pull request for the option critic branch, so far there is the network builder and preliminary agent

### CODE REVIEW

**Comment by peasant98** [comment]
Can you add a brief description to the PR?

**Comment by watakandai** [comment]
can you push the most recent update? I finally had time to go through the paper, code and your code. I now have an idea of how it should be implemented. Hope I can help.

File: pfrl/agents/option_critic.py [hunk 1]
```diff
@@ -189,7 +211,29 @@ def actor_loss_fn(self, obs, option, logp, entropy, reward, done, next_obs):
...
+        next_Q_prime = self.oc_prime.get_Q(next_states_prime)
+
+        next_states = self.oc.get_state(self.to_tensor(next_obs)).squeeze(0)
+        next_termination_probs = self.oc.get_terminations(next_states).detach()
+        next_options_term_prob = next_termination_probs[batch_idx, options]
+
+        gt = rewards + masks * self.gamma * ((1-next_options_term_prob) * next_Q_prime[batch_idx, options] \
+                                             + next_options_term_prob * next_Q_prime.max(dim=-1)[0])
+
+        td_err = (Q[batch_idx, options] - gt.detach()).pow(2).mul(0.5).mean()
```
Comment (watakandai) [inline pfrl/agents/option_critic.py:248]: maybe us MSELoss here? 
https://pytorch.org/docs/stable/generated/torch.nn.MSELoss.html

File: pfrl/agents/option_critic.py [hunk 2]
```diff
@@ -189,7 +211,29 @@ def actor_loss_fn(self, obs, option, logp, entropy, reward, done, next_obs):
...
+        Q = self.oc.get_Q(states)
+
+        next_states_prime = self.oc_prime.get_state(self.to_tensor(next_obs)).squeeze(0)
+        next_Q_prime = self.oc_prime.get_Q(next_states_prime)
+
+        next_states = self.oc.get_state(self.to_tensor(next_obs)).squeeze(0)
+        next_termination_probs = self.oc.get_terminations(next_states).detach()
+        next_options_term_prob = next_termination_probs[batch_idx, options]
+
+        gt = rewards + masks * self.gamma * ((1-next_options_term_prob) * next_Q_prime[batch_idx, options] \
```
Comment (watakandai) [inline pfrl/agents/option_critic.py:245]: Add comments to the important equations and refer to the equation numbers in the paper.

File: pfrl/agents/option_critic.py (hunk 1 above)
Comment (watakandai) [reply pfrl/agents/option_critic.py:248]: I skimmed through and the math seemed to be all correct. How is the training result?

**Comment by watakandai** [comment]
For `OC` class, add `AttributeSavingMixin` as one of the inheriting classes. 
https://github.com/HIRO-group/pfrl/blob/d1a94a871c1456a600dadd17b6c8af76d61d2207/pfrl/agents/option_critic.py#L97

`AttributeSavingMixin` is located in `agent.py` in pfrl. It provides the class an easy way to save/load parameters.

Add attributes to the class like below to select which attributes to store. 
For further info, check the `TD3` Agent. https://github.com/HIRO-group/pfrl/blob/master/pfrl/agents/td3.py
```
saved_attributes = (
        "policy",
        "q_func1",
        "q_func2",
        "target_policy",
        "target_q_func1",
        "target_q_func2",
        "policy_optimizer",
        "q_func1_optimizer",
        "q_func2_optimizer",
    )
```

File: examples/optioncritic/test_oc.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        num_eps = 1000
+
+        feature_output_size = 32
+        state_size = 4
+        num_options = 2
+        num_actions = 2
+        featureNetwork = nn.Linear(state_size, feature_output_size)
+        QNetwork = nn.Linear(feature_output_size, num_options)
+        terminationNetwork = nn.Linear(feature_output_size, num_options)
+        network = OptionCriticNetwork(featureNetwork, terminationNetwork, QNetwork, feature_output_size, num_options, num_actions)
```
Comment (watakandai) [inline examples/optioncritic/test_oc.py]: it would be better if it could automatically select gpu or cpu.

File: pfrl/agents/option_critic.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        self.eps_start = eps_start
+        self.eps_min = eps_min
+        self.eps_decay = eps_decay
+        self.num_steps = 0
+
+        self.features = featureNetwork
+        self.terminations = terminationNetwork
+        self.Q = QNetwork
+        self.options_W = nn.Parameter(torch.zeros(num_options, feature_output_size, num_actions))
+        self.options_b = nn.Parameter(torch.zeros(num_options, num_actions))
```
Comment (watakandai) [inline pfrl/agents/option_critic.py:45]: why don't you use `nn.Linear` for `num_options`?

File: pfrl/agents/option_critic.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+        self.device = device
+
+        self.to(device)
+
+    def get_state(self, obs):
```
Comment (watakandai) [inline pfrl/agents/option_critic.py]: `convert_to_state` makes more sense.

File: pfrl/agents/option_critic.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    def get_Q(self, state):
+        """Get Q value from state."""
+        return self.Q(state)
+
+    def predict_option_termination(self, state, current_option):
+        """Predict whether option will terminate, return (termination, next option)."""
+        termination = self.terminations(state)[current_option].sigmoid()
+        option_termination = Bernoulli(termination).sample()
+
+        return bool(option_termination.item())
```
Comment (watakandai) [inline pfrl/agents/option_critic.py:66]: Was this mentioned somewhere in the paper? I just want to know where this is coming from.

File: pfrl/agents/option_critic.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        action_dist = logits.softmax(dim=-1)
+        action_dist = Categorical(action_dist)
+
+        action = action_dist.sample()
+        logp = action_dist.log_prob(action)
+        entropy = action_dist.entropy()
+
+        return action.item(), logp, entropy
+
+    def greedy_option(self, state):
```
Comment (watakandai) [inline pfrl/agents/option_critic.py]: `get_greedy_option`

File: pfrl/agents/option_critic.py (hunk 6 above)
Comment (emeldar) [reply pfrl/agents/option_critic.py:66]: The use of sigmoid functions on the end of the termination functions is in the experiments section of the paper. Each termination function returns a probability of terminating given a state and current option. We then use the Bernoulli distribution to pick whether it ends or not using the probability generated.

File: pfrl/agents/option_critic.py (hunk 1 above)
Comment (emeldar) [reply pfrl/agents/option_critic.py:248]: training_graph

Comment (emeldar) [reply pfrl/agents/option_critic.py:248]: It's... interesting

Comment (watakandai) [reply pfrl/agents/option_critic.py:248]: this is very interesting .... lol

Comment (watakandai) [reply pfrl/agents/option_critic.py:248]: did you render how the agent was performing?

**Comment by peasant98** [comment]
what's the status of this -- should it be closed?

**Comment by emeldar** [comment]
Yes, should be closed
